### ISSUE TITLE
Separate metadata from the actual response in async search index

### ISSUE THREAD

**Issue by jimczi** [elastic/elasticsearch#71223]
Getting the [status of an async search](https://www.elastic.co/guide/en/elasticsearch/reference/current/async-search.html#get-async-search-status) is fast when the query is running but can be very costly if the response is already stored. 
When the task is gone/finished, we interrogate the async search index to retrieve the full response if it is still available. The metadata and statistics are stored inside a binary field that also contains the actual response so the cost to retrieve the status depends greatly on the size of the search response. 
We should separate the actual response from the metadata in order to allow to query these informations separately.
That would make the cost of status calls more constant and cheaper than retrieving the full response each time.

**Comment by mayya-sharipova** [comment]
Currently _async_search/status for completed searches returns:

for successfully completed searches:
```js
{
  "id" : "FmRldE8zREVEUzA2ZVpUeGs2ejJFUFEaMkZ5QTVrSTZSaVN3WlNFVmtlWHJsdzoxMDc=",
  "is_running" : false,
  "is_partial" : false,
  "start_time_in_millis" : 1583945890986,
  "expiration_time_in_millis" : 1584377890986,
  "_shards" : {
      "total" : 562,
      "successful" : 562,
      "skipped" : 0,
      "failed" : 0
  },
 "completion_status" : 200 
}
```

for unsuccessfully completed searches:
```
{
  "id" : "FmRldE8zREVEUzA2ZVpUeGs2ejJFUFEaMkZ5QTVrSTZSaVN3WlNFVmtlWHJsdzoxMDc=",
  "is_running" : false,
  "is_partial" : true,
  "start_time_in_millis" : 1583945890986,
  "expiration_time_in_millis" : 1584377890986,
  "_shards" : {
      "total" : 562,
      "successful" : 450,
      "skipped" : 0,
      "failed" : 112
  },
 "completion_status" : 503 
}
```

**Comment by mayya-sharipova** [comment]
## Proposal: Add extra fields 

```js
"properties" : {
  "expiration_time" : {
    "type" : "long"
  },
  "headers" : {
    "type" : "object",
    "enabled" : false
  },
  "response_headers" : {
    "type" : "object",
    "enabled" : false
  },
  "result" : {
    "type" : "object",
    "enabled" : false
  }
}
```

**Comment by mayya-sharipova** [comment]
We had a team discussion and decided to with the  Proposal 1. So the plan is:

- Introduce `status` as a single `binary` field in existing `.async-search` indices
- `status` can be either:
  -  a stored only  field
  - a doc values only field
- Do performance benchmarking with around 1000 docs to compare stored VS doc values fields performances
  -  some docs should have huge responses
  - 1000 requests per second, per 10 seconds

**Comment by dnhatn** [comment]
> Older docs indexed in  _async_search don't contain the status stored field, and a GET request with stored_fields will have stored fields empty (or GET request will return unfound if we are trying to find older docs in a new index). In this case, we need to reissue another GET request with _source (or older index).

@mayya-sharipova I think we can add a version to `AsyncExecutionId`(and encoded id) then make a decision based on the version so we can avoid issue two requests in a mixed cluster.

**Comment by mayya-sharipova** [comment]
@dnhatn Thanks, that's a great proposal. I will experiment with this idea

**Comment by mayya-sharipova** [comment]
We had another discussion on this topic today, and decided to go with GET response, and for a search response we need to worry about refresh interval.

We need to measure performance of retrieving  doc values fields VS stored field fields for status, and if retrieving doc values fields turns out to be much faster than stored fields, we can think of adding support for retrieving doc value fields to GET api.

**Comment by mayya-sharipova** [comment]
I've benchmarked the time needed for retrieving stored fields through GET request VS  retrieving doc values fields through SEARCH request depending on the size of `result` field. Here are the results:

number of docs in the index: 1000

| Result size | GET time ms | SEARCH time ms       | SEARCH time for first request ms |
| ----------- | ----------: | ----------------------------: | ----------------------------: |
| 100Kb       | 1.1         | 1.9 |6.2|
| 1Mb         | 2.4 | 3.0 |6.0|
| 10Mb |             |                               ||

**Comment by jimczi** [comment]
Are you sure that you're benchmarking the doc values case with `"stored_fields": ["_none_"]` ? I guess that could explain why responses with doc values are not fast. `100ms` for a status query is quite a lot, especially when testing in isolation. Is it only on the first retrieval or it's consistently slow ?

**Comment by mayya-sharipova** [comment]
> Are you sure that you're benchmarking the doc values case with "stored_fields": ["_none_"] ?

Great question! No, I only disabled _source, _id field was still being read from stored field.  Doing ` "stored_fields": "_none_"`, I've got much better benchmarks for doc_values:


| Result size | GET time with stored fields, ms | SEARCH time with doc values, ms       | SEARCH time with doc values for first request, ms |
| ----------- | ----------: | ----------------------------: | ----------------------------: |
| 10Mb |           17.7  |                               1.9| 7.8|

**Comment by jimczi** [comment]
Considering that response in the `MBs` range are more anomalies than realistic response sizes, I would still go with the stored field option. It was made for that purpose so it's logical to use it here. My rough feeling is that big responses are rather in a range between 100k and 1MB so the number shared above for stored fields are good enough. Do you have the numbers for the current full source solution in that range ?

**Comment by jtibshirani** [comment]
I'd also be really curious to see a comparison against loading from _source in the 100KB - 1MB range. I tried a similar experiment on metricbeat data and only saw a small improvement when moving from _source to stored fields: https://github.com/elastic/elasticsearch/issues/9034#issuecomment-714092754. This made me think that JSON parsing overhead is low. However the metricbeat documents are substantially smaller than 1MB, it would be helpful to see this datapoint as well.

**Comment by mayya-sharipova** [comment]
Index contains 1000 docs
10 docs: 1Mb
990 docs:  100Kb

|Doc size | GET _source |  GET stored fields | SEARCH doc values |
| ----------- | ----------: | ----------------------------: | ----------------------------: |
| 100 kb |           1.3 ms  |                               1.3 ms|  1.4 ms|
|  1 Mb |          4.7 ms  |                               3.0 ms| 1.4 ms|

**Comment by mayya-sharipova** [comment]
We've discussed this again, and considering that `_async_search` index should  NOT include big responses,  and for responses < 1Mb, retrieving `_source` or a separate stored field doesn't make much difference, we've decided for now NOT to proceed with separating metadata from the actual response. 

But we still would like to keep the issue open, as there could be other ways to improve retrieving metadata:
 -  Currently the whole _source is getting sent from a coordinating node to a data node. Can we send  to a coordinating node only `status` (metadata)? Perhaps if we know that first X bytes of _source constitute status, can we send only first X bytes?
 -  Can `status` (metadata) be a binary header?  
 - Can we add a circuit breaker when we request too big of a source to load? 
- Related to this, an [issue to break up huge async search response into several docs](https://github.com/elastic/elasticsearch/issues/67594)? Perhaps, `status` can be stored only in the 1st of this docs?
